tracing-subscriber/src/registry/sharded.rs

```diff
@@ -270,3 +320,100 @@ impl<'a> SpanData<'a> for Data<'a> {
         ExtensionsMut::new(self.inner.extensions.write().expect("Mutex poisoned"))
     }
 }
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::Registry;
+    use crate::{layer::Context, registry::LookupSpan, Layer};
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use tracing::{self, subscriber::with_default};
+    use tracing_core::{
+        span::{Attributes, Id},
+        Subscriber,
+    };
+
+    struct NopLayer;
+    impl<S> Layer<S> for NopLayer
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+            assert!(&ctx.span(&id).is_some());
+        }
+    }
+
+    struct NopLayer2;
+    impl<S> Layer<S> for NopLayer2
+    where
+        S: Subscriber + for<'a> LookupSpan<'a>,
+    {
+        fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+            assert!(&ctx.span(&id).is_some());
+        }
+    }
+
+    #[test]
+    fn single_layer_can_access_closed_span() {
+        let subscriber = NopLayer
+            .and_then(NopLayer)
+            .with_subscriber(Registry::default());
+
+        with_default(subscriber, || {
+            let span = tracing::debug_span!("span");
+            drop(span);
+        });
+    }
+
+    #[test]
+    fn multiple_layers_can_access_closed_span() {
+        let subscriber = NopLayer
+            .and_then(NopLayer)
+            .and_then(NopLayer2)
```

Comment:
Nit/TIOLI: two separate layer types isn't _strictly_ necessary, you could just use a whole bunch of layers with the same type...not a big deal, and I see why we _might_ want to test that different layers of different types also see the right behavior, but there would be less boilerplate.

Reply:
hrm, fair point. They would just be called `n` times, where `n` is the amount of times we compose those layers.